lib/binary_to_decimal.rb

```diff
@@ -4,6 +4,51 @@
 # The least significant bit is at index 7.
 # Calculate  and return the decimal value for this binary number using
 # the algorithm you devised in class.
+
+require 'pry'
+
 def binary_to_decimal(binary_array)
-  raise NotImplementedError
+  if binary_array.class != Array
+    raise ArgumentError, 'Input is not an array.'
+  end
+  
+  decimal = 0
+  
+  binary_array.each_with_index do |bit, i|
+    if bit == 1
+      decimal += 2**(7-i)
+    end
+  end
+  
+  return decimal
+end
+
+# Determines the length of a binary array given a decimal number.
+def find_length_of_binary(num)
+  cur_exponent = 0
+  
+  # while the current power of 2 is not large enough to be greater than or equal to the decimal
+  while 2**cur_exponent <= num
+    # increase the exponent by 1
+    cur_exponent += 1
+  end
+  
+  return cur_exponent
+end
+
+# Returns an array representing the binary form of a decimal number. 
+def decimal_to_binary(num)
+  length_binary = find_length_of_binary(num)
+  binary_array = []
+  
+  length_binary.times do |i|
+    if num - 2**(length_binary-i-1) >= 0
+      binary_array << 1
+      num -= 2**(length_binary-i-1)
```

Comment:
You're doing this subtraction twice, but here and in the conditional.  Could you dry up the code a bit?  Maybe do it right after the loop starts and just compare `if num >= 0`